src/charm.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical
+# See LICENSE file for licensing details.
+
+"""Parca Scrape Target Charm."""
+
+import logging
+import ssl
+from typing import Dict, List, Literal, Optional, TypedDict
+from urllib.parse import urlparse
+
+import ops
+from charms.parca_k8s.v0.parca_scrape import ProfilingEndpointProvider
+
+logger = logging.getLogger(__name__)
+
+ScrapeJob = Dict[str, List[str]]
+
+
+class TLSConfig(TypedDict, total=False):
+    """TLS config type."""
+
+    insecure_skip_verify: bool
+    ca: str
+    server_name: str
+
+
+class ScrapeJobsConfig(TypedDict, total=False):
+    """Scrape job config type."""
+
+    static_configs: List[ScrapeJob]
+    scheme: Optional[Literal["https", "http"]]
+    tls_config: TLSConfig
+
+
+class ParcaScrapeTargetCharm(ops.CharmBase):
+    """Parca Scrape Target Charm."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        # ENDPOINT WRAPPERS
+        self._profiling = ProfilingEndpointProvider(self, jobs=self._scrape_jobs)
+
+        # event handlers
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_unit_status)
+
+        # unconditional logic
+        self._reconcile()
+
+    # RECONCILERS
+    def _reconcile(self):
+        """Unconditional logic to run regardless of the event we're processing."""
+        self._reconcile_relations()
+
+    def _reconcile_relations(self):
+        self._profiling.set_scrape_job_spec()
+
+    # SCRAPE JOB PROPERTIES
+    @property
+    def _scrape_jobs(self) -> Optional[List[ScrapeJobsConfig]]:
+        """Set up Parca scrape configuration for external targets."""
+        # return None if no targets are configured
+        if not self._targets:
+            return None
+
+        job: ScrapeJobsConfig = {
+            "static_configs": [{"targets": self._targets}],
+        }
+        if self._scheme == "https":
+            job["scheme"] = "https"
+            job["tls_config"] = self._tls_config
+
+        return [job]
+
+    # CONFIG PROPERTIES
+    @property
+    def _tls_config(self) -> TLSConfig:
+        """Get the TLS configuration from the Juju model config."""
+        tls_config: TLSConfig = {
+            "insecure_skip_verify": self._tls_insecure_skip_verify,
+        }
+        if ca := self._tls_ca_cert:
+            tls_config["ca"] = ca
+        if server_name := self._tls_server_name:
+            tls_config["server_name"] = server_name
```

Comment:
Here, when filling in the `scheme` and `tls_config`, the values of `scheme`, `tls_ca_cert`, and `tls_server_name` are not validated. This means that invalid data might be passed to the integration, especially for `tls_ca_cert`, and `tls_server_name`. Perhaps we should validate the input values first and skip the update if they are invalid?